### ISSUE TITLE
Numbers skip frequently and don't follow an order.

### ISSUE THREAD

**Issue by tiancola** [mikel-brostrom/Yolov5_DeepSort_Pytorch#298]
Thank you for your great work.
I have a question--why the id are assigned without any order?
I think some object is detected and given an ID, but it is not tracked, so some IDs are not displayed.
I am not sure why this is happening. Any help is much appreciated.Thanks.

**Comment by songjiahao-wq** [comment]
Did you solve this problem?

**Comment by ZiyueWangUoB** [comment]
That is how deepsort and sort works. There is a ```n_init``` variable in the tracker.py file within deepsort/sort. Each track is created after the initial detection, but needs a number of continuous detections for it to be initialised. Only after it is initialised will the track be displayed. 


If you want all the IDs to show, then you can change this value here https://github.com/mikel-brostrom/Yolov5_DeepSort_Pytorch/blob/8bd72cadfae822bbabdc67a45b9d57abbfdf0135/track.py#L55
to 0, which would show all the IDs. However this will be very messy.

**Comment by HeChengHui** [comment]
@ZiyueWangUoB 
Correct me if I am wrong. Does that mean that every object will be given an ID when it was first detected instead of given an ID after appearing for at least 3 frames?

**Comment by ZiyueWangUoB** [comment]
@HeChengHui No, every new object will be given an ID at first detection. The waiting 3 frames is only for counting the track as "alive" and hence showing it. By changing `n_init` to 0, all tracks would be "alive" at the first detection, which would be messy but sometimes necessary.

**Comment by HeChengHui** [comment]
@ZiyueWangUoB 
So let's say 2 objects are detected. ID 1 and 2 will be given to each. If object 2 does not show for the next 2 frames (to get 3 consecutive hits, aka "alive"), it won't be shown on the screen. 
Now, another object detected is detected for the next 3 frames. So that object will be shown with ID of 3?

**Comment by ZiyueWangUoB** [comment]
@HeChengHui yes that's exactly how it works

**Comment by HeChengHui** [comment]
@ZiyueWangUoB 
Thank you!

If I want only "alive" objects to be given an ID instead of every detected, is there a way to do this?

**Comment by ZiyueWangUoB** [comment]
@HeChengHui There is no built in way. Your best bet is to use a `dict` to store the alive objects and map a new ID to those. Then overwrite the IDs being displayed and saved with your own mapped IDs.

**Comment by HeChengHui** [comment]
@ZiyueWangUoB That's a shame. Do you happen to know which file should I look into for that?

**Comment by ZiyueWangUoB** [comment]
@HeChengHui Create your own `dict` and change the id here 
https://github.com/mikel-brostrom/Yolov5_DeepSort_Pytorch/blob/5af8ecc7aa3cfe153273b709abb22a214acffebd/track.py#L172
to the one you've mapped yourself

**Comment by HeChengHui** [comment]
@ZiyueWangUoB 
Alright, thanks for the help!

**Comment by Ultiva** [comment]
> @ZiyueWangUoB
> Alright, thanks for the help!

Were you able to improve adapt it for your purposes ?

**Comment by HeChengHui** [comment]
@Ultiva 
Yes. I used the old id as the key and a value starting from 0. whenever a new id is found, the value increment by 1.

**Comment by songjiahao-wq** [comment]
> @Ultiva 是的。我使用旧的id作为键和从0开始的值。每当找到新 ID 时，该值都会递增 1。
请问能否告知一下如何更改？

**Comment by HeChengHui** [comment]
@songjiahao-wq 

```
 ## changed the way id is calculated (using dict)
                        id = output[4]
                        if id in new_id_dict.keys():
                            id = new_id_dict[id]
                        else:
                            new_id_dict[id] = new_id_index
                            id = new_id_index
                            new_id_index += 1
```

**Comment by songjiahao-wq** [comment]
> @songjiahao-wq
> 
> ```
>  ## changed the way id is calculated (using dict)
>                         id = output[4]
>                         if id in new_id_dict.keys():
>                             id = new_id_dict[id]
>                         else:
>                             new_id_dict[id] = new_id_index
>                             id = new_id_index
>                             new_id_index += 1
> ```
> 
> new_id_index is a int value. Start from 0.

非常感谢

**Comment by Ultiva** [comment]
Looks good. I implemented in the meanwhile a slightly different approach... I introduced a new attribute for track  is called visable_id and chenged id to track_id. Track id stays basically the same as before and visable_id just gets assigned and increased once a track gets confirmed (gets actually visable). Works also fine and I have an overview how much tracks are started with reference to the actually visable_ids.
